GitHub: https://github.com/applecrumble123/SIT799_Human_Aligned_Artificial_Intelligence

Welcome to your assignment this week! 

To better understand bias and discrimination in AI, in this assignment, we will look at a Natural Language Processing use case. 

Natural Language Pocessing (NLP) is a branch of Artificial Intelligence (AI) that helps computers to understand, to interpret and to manipulate natural (i.e. human) language.
Imagine NLP-powered machines as black boxes that are capable of understanding and evaluating the context of the input documents (i.e. collection of words), outputting meaningful results that depend on the task the machine is designed for.


![](images/1_3zMvUnPzYZF9CSHdj6hT5A.png "the title")



<caption><center> Documents are fed into magic NLP model capable to get, for instance, the sentiment of the original content</center></caption>



Just like any other machine learning algorithm, biased data results in biased outcomes. And just like any other algorithm, results debiasing is painfully annoying, to the point that it might be simpler to unbias the society itself.

# The big deal: word embeddings

Words must be represented as **numeric vectors** in order to be fed into machine learning algorithms. One of the most powerful (and popular) ways to do it is through **Word Embeddings**. In word embedding models, each word in a given language is assigned to a high-dimensional vector, such that **the geometry of the vectors captures relations between the words**. 


Because word embeddings are very computionally expensive to train, most ML practitioners will load a pre-trained set of embeddings. 

**After this assignment you will be able to:**

- Load pre-trained word vectors, and measure similarity using cosine similarity
- Use word embeddings to solve word analogy problems such as Man is to Woman as King is to ______. 
- Modify word embeddings to reduce their gender bias 

Let's get started! Run the following cell to install all the packages you will need.

In [1]:
#!pip install numpy
#!pip install keras
#!pip install tensorflow

Run the following cell to load the packages you will need.

In [2]:
import numpy as np
from w2v_utils import *

Using TensorFlow backend.


Next, lets load the word vectors. For this assignment, we will use 50-dimensional GloVe vectors to represent words. Run the following cell to load the `word_to_vec_map`. 

In [3]:
words, word_to_vec_map = read_glove_vecs('data/glove.6B.50d.txt')

You've loaded:
- `words`: set of words in the vocabulary.
- `word_to_vec_map`: dictionary mapping words to their GloVe vector representation.



In [4]:
list_of_words = list(words)
print("Example of words: ",list(words)[:10])
print("Vector for word 'person' = ", word_to_vec_map.get('person'))



Example of words:  ['wavefront', 'landquart', 'wyeth', 'goetschl', 'capillaries', 'naby', 'setapak', 'minstrels', 'lvds', 'anko']
Vector for word 'person' =  [ 0.61734    0.40035    0.067786  -0.34263    2.0647     0.60844
  0.32558    0.3869     0.36906    0.16553    0.0065053 -0.075674
  0.57099    0.17314    1.0142    -0.49581   -0.38152    0.49255
 -0.16737   -0.33948   -0.44405    0.77543    0.20935    0.6007
  0.86649   -1.8923    -0.37901   -0.28044    0.64214   -0.23549
  2.9358    -0.086004  -0.14327   -0.50161    0.25291   -0.065446
  0.60768    0.13984    0.018135  -0.34877    0.039985   0.07943
  0.39318    1.0562    -0.23624   -0.4194    -0.35332   -0.15234
  0.62158    0.79257  ]


GloVe vectors provide much more useful information about the meaning of individual words. Lets now see how you can use GloVe vectors to decide how similar two words are. 


# Cosine similarity

To measure how similar two words are, we need a way to measure the degree of similarity between two embedding vectors for the two words. Given two vectors $u$ and $v$, cosine similarity is defined as follows: 

\begin{equation}
\text{CosineSimilarity(u, v)} = \frac {u . v} {||u||_2 ||v||_2} = cos(\theta) \tag{1}
\end{equation}

where $u.v$ is the dot product (or inner product) of two vectors, $||u||_2$ is the norm (or length) of the vector $u$, and $\theta$ is the angle between $u$ and $v$. This similarity depends on the angle between $u$ and $v$. If $u$ and $v$ are very similar, their cosine similarity will be close to 1; if they are dissimilar, the cosine similarity will take a smaller value. 


![](images/cosine_sim.png)

<caption><center> <b>Figure 1</b>: The cosine of the angle between two vectors is a measure of how similar they are</center></caption>



***

**Task 1**: Implement the function `cosine_similarity()` to evaluate similarity between word vectors.

**Reminder**: The norm of $u$ is defined as $||u||_2 = \sqrt{\sum_{i=1}^{n} u_i^2}$

In [5]:
## Task 1
# cosine_similarity

import math 

def cosine_similarity_function(u, v):
    
    dot_product_u_v = np.dot(u,v)
    
    sum_u = 0
    for i in u:
        i = i ** 2 
        sum_u = sum_u + i
        
    norm_2_u = math.sqrt(sum_u)
    
    sum_v = 0
    for j in v:
        j = j ** 2 
        sum_v = sum_v + j
        
    norm_2_v = math.sqrt(sum_v)
    
    
    cosine_similarity = dot_product_u_v/(norm_2_u*norm_2_v)
    
    return cosine_similarity
    

    ## START YOU CODE HERE
    
similarity1 = cosine_similarity_function(word_to_vec_map.get(list_of_words[0]),  word_to_vec_map.get(list_of_words[1]))
print("Similarity between '{}' and '{}' is {}".format(list_of_words[0], list_of_words[1], similarity1))

similarity2 = cosine_similarity_function(word_to_vec_map.get(list_of_words[10]),  word_to_vec_map.get(list_of_words[20]))
print("Similarity between '{}' and '{}' is {}".format(list_of_words[10], list_of_words[20], similarity2))

    ## END

Similarity between 'wavefront' and 'landquart' is 0.009691132500844912
Similarity between 'hele' and 'kobo' is 0.44207394947399953


***

**Task 2**: Implement `most_similar_word` which returns the most similar word to a word.

In [6]:
## Task 2
# GRADED FUNCTION: most_similar_word

def most_similar_word(word, word_to_vec):
    
    """
    Most similar word return the most similar word to the word u.
        
    Arguments:
        word -- a word, string        
        word_to_vec_map -- dictionary that maps words to their corresponding vectors. 
    
    Returns:
    best_word --  the most similar word to u as measured by cosine similarity
    """ 
    
    

    v = word_to_vec_map.get(word)
    
    sum_v = 0
    for k in v:
        k_square = k ** 2 
        sum_v = sum_v + k_square

    norm_2_v = math.sqrt(sum_v)
    
    # duplicate the list
    copy_list_of_words = list_of_words.copy()
    
    # returns a value without 'u' in the list
    new_list_of_words = [i for i in copy_list_of_words if i != 'u']
    
    array_cosine_similarity = []
    
    for i in new_list_of_words:
        
        u = word_to_vec_map.get(i)
        dot_product_i_word = np.dot(u,v)
    
        sum_u = 0
        for j in u:
            j_square = j ** 2 
            sum_u = sum_u + j_square

        norm_2_u = math.sqrt(sum_u)


        cosine_similarity_u_v = dot_product_i_word/(norm_2_u*norm_2_v)
        
        array_cosine_similarity.append(cosine_similarity_u_v)
        
    
    # find max value
    max = array_cosine_similarity[0]
    for i in array_cosine_similarity:
        if i > max:
            max = i
    
    index = array_cosine_similarity.index(max)
    #print(index)
    
    highest_similarity_word = new_list_of_words[index]
    print("The most similar word to 'u' in the list of words is '{}' with the cosine similarity of {}".format(
        highest_similarity_word, max))

    
    ## START YOU CODE HERE
most_similar_word('u', word_to_vec_map.get('u'))

    ## END
 


The most similar word to 'u' in the list of words is 'o' with the cosine similarity of 0.7432858003299065


Answer the questions below:

**TASK 3: Write a code the answer the following questions:**

***

What is the similarity between the words brother and friend?

In [7]:
brother_vector = word_to_vec_map.get('brother')


friend_vector = word_to_vec_map.get('friend')

dot_product_brother_friend = np.dot(brother_vector,friend_vector)

sum_brother = 0
for i in brother_vector:
    i = i ** 2 
    sum_brother = sum_brother + i

norm_2_brother_vector = math.sqrt(sum_brother)

sum_friend = 0
for j in friend_vector:
    j = j ** 2 
    sum_friend = sum_friend + j

norm_2_friend_vector = math.sqrt(sum_friend)


cosine_similarity_brother_friend = dot_product_brother_friend/(norm_2_brother_vector*norm_2_friend_vector)
print("The similarity between the words 'brother' and 'friend' is {}".format(cosine_similarity_brother_friend))


The similarity between the words 'brother' and 'friend' is 0.8713178668124655


What is the similarity between the words computer and kid?

In [8]:
computer_vector = word_to_vec_map.get('computer')

kid_vector = word_to_vec_map.get('kid')


dot_product_computer_kid = np.dot(computer_vector,kid_vector)

sum_computer = 0
for i in computer_vector:
    i = i ** 2 
    sum_computer = sum_computer + i

norm_2_computer_vector = math.sqrt(sum_computer)


sum_kid = 0
for j in kid_vector:
    j = j ** 2 
    sum_kid = sum_kid + j

norm_2_kid_vector = math.sqrt(sum_kid)


cosine_similarity_computer_kid = dot_product_computer_kid/(norm_2_computer_vector*norm_2_kid_vector)
print("The similarity between the words 'computer' and 'kid' is {}".format(cosine_similarity_computer_kid))



The similarity between the words 'computer' and 'kid' is 0.43800166210363856


What is the similarity between the words V1=(france - paris) and V2=(rome - italy)?

In [9]:
V1 = word_to_vec_map.get('france') - word_to_vec_map.get('paris')
V2 = word_to_vec_map.get('rome') - word_to_vec_map.get('italy')

dot_product_V1_V2 = np.dot(V1,V2)

sum_V1 = 0
for i in V1:
    i = i ** 2 
    sum_V1 = sum_V1 + i

norm_2_V1 = math.sqrt(sum_V1)


sum_V2 = 0
for j in V2:
    j = j ** 2 
    sum_V2 = sum_V2 + j

norm_2_V2 = math.sqrt(sum_V2)


cosine_similarity_V1_V2 = dot_product_V1_V2/(norm_2_V1*norm_2_V2)
print("The similarity between the words 'V1 (france - paris)' and 'V2 (rome - italy)' is {}".format(cosine_similarity_V1_V2))


The similarity between the words 'V1 (france - paris)' and 'V2 (rome - italy)' is -0.6751479308174203


What is the most similar word to computer?


In [10]:
v = word_to_vec_map.get('computer')
    
sum_v = 0
for k in v:
    k_square = k ** 2 
    sum_v = sum_v + k_square

norm_2_v = math.sqrt(sum_v)

# duplicate the list
copy_list_of_words = list_of_words.copy()

# returns a value without 'computer' in the list
new_list_of_words = [i for i in copy_list_of_words if i != 'computer']

array_cosine_similarity = []

for i in new_list_of_words:

    u = word_to_vec_map.get(i)
    dot_product_i_word = np.dot(u,v)

    sum_u = 0
    for j in u:
        j_square = j ** 2 
        sum_u = sum_u + j_square

    norm_2_u = math.sqrt(sum_u)


    cosine_similarity_u_v = dot_product_i_word/(norm_2_u*norm_2_v)

    array_cosine_similarity.append(cosine_similarity_u_v)


# find max value
max = array_cosine_similarity[0]
for i in array_cosine_similarity:
    if i > max:
        max = i

index = array_cosine_similarity.index(max)
#print(index)

highest_similarity_word = new_list_of_words[index]
print("The most similar word to 'computer' in the list of words is '{}' with the cosine similarity of {}".format(
    highest_similarity_word, max))

The most similar word to 'computer' in the list of words is 'computers' with the cosine similarity of 0.9165044765653498


What is the most similar word to australia?

In [11]:
v = word_to_vec_map.get('australia')
    
sum_v = 0
for k in v:
    k_square = k ** 2 
    sum_v = sum_v + k_square

norm_2_v = math.sqrt(sum_v)

# duplicate the list
copy_list_of_words = list_of_words.copy()

# returns a value without 'australia' in the list
new_list_of_words = [i for i in copy_list_of_words if i != 'australia']

array_cosine_similarity = []

for i in new_list_of_words:

    u = word_to_vec_map.get(i)
    dot_product_i_word = np.dot(u,v)

    sum_u = 0
    for j in u:
        j_square = j ** 2 
        sum_u = sum_u + j_square

    norm_2_u = math.sqrt(sum_u)

    cosine_similarity_u_v = dot_product_i_word/(norm_2_u*norm_2_v)

    array_cosine_similarity.append(cosine_similarity_u_v)


# find max value
max = array_cosine_similarity[0]
for i in array_cosine_similarity:
    if i > max:
        max = i

index = array_cosine_similarity.index(max)
#print(index)

highest_similarity_word = new_list_of_words[index]
print("The most similar word to 'australia' in the list of words is '{}' with the cosine similarity of {}".format(
    highest_similarity_word, max))

The most similar word to 'australia' in the list of words is 'zealand' with the cosine similarity of 0.92077950622842


What is the most similar word to python?


In [12]:
v = word_to_vec_map.get('python')
    
sum_v = 0
for k in v:
    k_square = k ** 2 
    sum_v = sum_v + k_square

norm_2_v = math.sqrt(sum_v)

# duplicate the list
copy_list_of_words = list_of_words.copy()

# returns a value without 'python' in the list
new_list_of_words = [i for i in copy_list_of_words if i != 'python']

array_cosine_similarity = []

for i in new_list_of_words:

    u = word_to_vec_map.get(i)
    dot_product_i_word = np.dot(u,v)

    sum_u = 0
    for j in u:
        j_square = j ** 2 
        sum_u = sum_u + j_square

    norm_2_u = math.sqrt(sum_u)

    cosine_similarity_u_v = dot_product_i_word/(norm_2_u*norm_2_v)

    array_cosine_similarity.append(cosine_similarity_u_v)


# find max value
max = array_cosine_similarity[0]
for i in array_cosine_similarity:
    if i > max:
        max = i

index = array_cosine_similarity.index(max)
#print(index)

highest_similarity_word = new_list_of_words[index]
print("The most similar word to 'python' in the list of words is '{}' with the cosine similarity of {}".format(
    highest_similarity_word, max))

The most similar word to 'python' in the list of words is 'reticulated' with the cosine similarity of 0.6916364915886599


***

Playing around the cosine similarity of other inputs will give you a better sense of how word vectors behave. 

#  Word analogy task

In the word analogy task, we complete the sentence <font color='brown'>"*a* is to *b* as *c* is to **____**"</font>. An example is <font color='brown'> '*man* is to *woman* as *king* is to *queen*' </font>. In detail, we are trying to find a word *d*, such that the associated word vectors $e_a, e_b, e_c, e_d$ are related in the following manner: $e_b - e_a \approx e_d - e_c$. We will measure the similarity between $e_b - e_a$ and $e_d - e_c$ using cosine similarity. 


**Task 4**: Complete the code below to be able to perform word analogies!

***

In [31]:
## Task 4
# GRADED FUNCTION: complete_analogy

import math

def complete_analogy(word_a, word_b, word_c, word_to_vec_map):
    """
    Performs the word analogy task as explained above: a is to b as c is to ____. 
    
    Arguments:
    word_a -- a word, string
    word_b -- a word, string
    word_c -- a word, string
    word_to_vec_map -- dictionary that maps words to their corresponding vectors. 
    
    Returns:
    best_word --  the word such that v_b - v_a is close to v_best_word - v_c, as measured by cosine similarity
    """
    
    a_vector = word_to_vec_map.get(word_a)
    b_vector = word_to_vec_map.get(word_b)
    
    v1 = b_vector - a_vector
    
    sum_V1 = 0
    for i in v1:
        i_square = i ** 2 
        sum_V1 = sum_V1 + i_square

    norm_2_V1 = math.sqrt(sum_V1)

    c_vector = word_to_vec_map.get(word_c)
    
    # copy a list of word
    copy_list_of_words = list_of_words.copy()

    # returns a value without word_c in the list
    new_list_of_words = [i for i in copy_list_of_words if i != word_c]
    
    
    array_cosine_similarity = []
    
    
    for i in new_list_of_words:

        new_word = word_to_vec_map.get(i)
        
        v2 = new_word - c_vector
        
        sum_V2 = 0
        for j in v2:
            j_square = j ** 2 
            sum_V2 = sum_V2 + j_square

        norm_2_V2 = math.sqrt(sum_V2)
        
        dot_product_v1_v2 = np.dot(v1,v2)
        cosine_similarity_u_v = dot_product_v1_v2/(norm_2_V1*norm_2_V2)
        
        array_cosine_similarity.append(cosine_similarity_u_v)
    
    # find max value
    max = array_cosine_similarity[0]
    for i in array_cosine_similarity:
        if i > max:
            max = i

    index = array_cosine_similarity.index(max)
    #print(index)

    highest_similarity_word = new_list_of_words[index]
    
    print("'{}' is to '{}' as '{}' is to '{}' of cosine similarity {}".format(
        word_a, word_b, word_c,highest_similarity_word, max))


        

    
    ## START YOU CODE HERE
    
complete_analogy(list_of_words[10], list_of_words[8], list_of_words[45], word_to_vec_map)

    ##ENFD


'hele' is to 'lvds' as 'politic' is to 'matroid' of cosine similarity 0.589543456472761


Run the cell below to test your code, this may take 1-2 minutes.

In [14]:
triads_to_try = [('italy', 'italian', 'spain'), ('india', 'delhi', 'japan'), ('man', 'woman', 'boy'), ('small', 'smaller', 'large')]
for triad in triads_to_try:
    print ('{} -> {} :: {} -> {}'.format( *triad, complete_analogy(*triad,word_to_vec_map, 0.1)))

italy and italian has the cosine similarity of 0.8423283703580938
spain and brazil has the cosine similarity of 0.8340149370596243
Therefore, italy is to italian as spain is to brazil

italy -> italian :: spain -> None
india and delhi has the cosine similarity of 0.8142766830734145
japan and korea has the cosine similarity of 0.8154973165622372
Therefore, india is to delhi as japan is to korea

india -> delhi :: japan -> None
man and woman has the cosine similarity of 0.886033771849582
boy and woman has the cosine similarity of 0.8596120447433488
Therefore, man is to woman as boy is to woman

man -> woman :: boy -> None
small and smaller has the cosine similarity of 0.8903936492649787
large and smaller has the cosine similarity of 0.8906458656865412
Therefore, small is to smaller as large is to smaller

small -> smaller :: large -> None


Once you get the correct expected output, please feel free to modify the input cells above to test your own analogies. Try to find some other analogy pairs that do work, but also find some where the algorithm doesn't give the right answer: For example, you can try small->smaller as big->?.  

# Debiasing word vectors

In the following exercise, you will examine gender biases that can be reflected in a word embedding, and explore algorithms for reducing the bias. In addition to learning about the topic of debiasing, this exercise will also help hone your intuition about what word vectors are doing. This section involves a bit of linear algebra, though you can probably complete it even without being expert in linear algebra, and we encourage you to give it a shot. 

Lets first see how the GloVe word embeddings relate to gender. You will first compute a vector $g = e_{woman}-e_{man}$, where $e_{woman}$ represents the word vector corresponding to the word *woman*, and $e_{man}$ corresponds to the word vector corresponding to the word *man*. The resulting vector $g$ roughly encodes the concept of "gender". (You might get a more accurate representation if you compute $g_1 = e_{mother}-e_{father}$, $g_2 = e_{girl}-e_{boy}$, etc. and average over them. But just using $e_{woman}-e_{man}$ will give good enough results for now.) 


**Task 5:** Compute the bias vector using woman - man

***

In [15]:
## START YOU CODE HERE

g = word_to_vec_map.get('woman') - word_to_vec_map.get('man')

## END

print(g)

[-0.087144    0.2182     -0.40986    -0.03922    -0.1032      0.94165
 -0.06042     0.32988     0.46144    -0.35962     0.31102    -0.86824
  0.96006     0.01073     0.24337     0.08193    -1.02722    -0.21122
  0.695044   -0.00222     0.29106     0.5053     -0.099454    0.40445
  0.30181     0.1355     -0.0606     -0.07131    -0.19245    -0.06115
 -0.3204      0.07165    -0.13337    -0.25068714 -0.14293    -0.224957
 -0.149       0.048882    0.12191    -0.27362    -0.165476   -0.20426
  0.54376    -0.271425   -0.10245    -0.32108     0.2516     -0.33455
 -0.04371     0.01258   ]


***

Now, you will consider the cosine similarity of different words with $g$. Consider what a positive value of similarity means vs a negative cosine similarity. 

**Task 6: Compute and print the similarity between g and the words in name_list**

***
    

In [16]:
print ('List of names and their similarities with constructed vector:')

name_list = ['john', 'marie', 'sophie', 'ronaldo', 'priya', 'rahul', 'danielle', 'reza', 'katy', 'yasmin']
print(name_list)

## START YOU CODE HERE

array_cosine_similarity = []

for i in name_list:
    
    name_vector = word_to_vec_map.get(i)


    dot_product_g_name_vector = np.dot(g,name_vector)

    sum_g = 0
    for i in g:
        i = i ** 2 
        sum_g = sum_g + i

    norm_2_g = math.sqrt(sum_g)


    sum_name_vector = 0
    for j in name_vector:
        j = j ** 2 
        sum_name_vector = sum_name_vector + j

    norm_2_name_vector = math.sqrt(sum_name_vector)


    cosine_similarity_g_name = dot_product_g_name_vector/(norm_2_g*norm_2_name_vector)
    
    array_cosine_similarity.append(cosine_similarity_g_name)

print(array_cosine_similarity)
    
    
    

## END 

List of names and their similarities with constructed vector:
['john', 'marie', 'sophie', 'ronaldo', 'priya', 'rahul', 'danielle', 'reza', 'katy', 'yasmin']
[-0.2316335614597372, 0.315597935396073, 0.3186878985941878, -0.3124479685032943, 0.17632041839009396, -0.16915471039231716, 0.24393299216283895, -0.07930429672199552, 0.2831068659572614, 0.23313857767928753]


***

**TASK 7: What do you observe?**

***

The female names have a positive cosine similarity and the male names have a negative cosine similarity.

***

**Task 8: Compute and print the similarity between g and the words in word_list:**

***

In [17]:
print('Other words and their similarities:')

word_list = ['lipstick', 'guns', 'science', 'arts', 'literature', 'warrior','doctor', 'tree', 'receptionist', 
             'technology',  'fashion', 'teacher', 'engineer', 'pilot', 'computer', 'singer']

print(word_list)

## START YOU CODE HERE

array_cosine_similarity = []

for i in word_list:
    
    word_vector = word_to_vec_map.get(i)


    dot_product_g_word_vector = np.dot(g,word_vector)

    sum_g = 0
    for i in g:
        i = i ** 2 
        sum_g = sum_g + i

    norm_2_g = math.sqrt(sum_g)


    sum_word_vector = 0
    for j in word_vector:
        j = j ** 2 
        sum_word_vector = sum_word_vector + j

    norm_2_word_vector = math.sqrt(sum_word_vector)


    cosine_similarity_g_word = dot_product_g_word_vector/(norm_2_g*norm_2_word_vector)
    
    array_cosine_similarity.append(cosine_similarity_g_word)

print(array_cosine_similarity)

## END 


Other words and their similarities:
['lipstick', 'guns', 'science', 'arts', 'literature', 'warrior', 'doctor', 'tree', 'receptionist', 'technology', 'fashion', 'teacher', 'engineer', 'pilot', 'computer', 'singer']
[0.27691916256382665, -0.1888485567898897, -0.060829065409296994, 0.008189312385880337, 0.0647250443345993, -0.20920164641125283, 0.11895289410935037, -0.07089399175478088, 0.3307794175059373, -0.13193732447554296, 0.03563894625772699, 0.1792092343182566, -0.0803928049452407, 0.0010764498991916935, -0.10330358873850497, 0.18500518136496286]


***

**TASK 9: What do you observe?**

***

"lipstick" and "receptionist" have the highest positive cosine similarity, showing a bias towards the female gender while "warrior" and "gun" have highest negative cosine similarity, showing a bias to the male gender. This observation is seen when being compared to the cosine similarity with respect to the female and male names.


***



We'll see below how to reduce the bias of these vectors, using an algorithm due to [Boliukbasi et al., 2016](https://arxiv.org/abs/1607.06520). Note that some word pairs such as "actor"/"actress" or "grandmother"/"grandfather" should remain gender specific, while other words such as "receptionist" or "technology" should be neutralized, i.e. not be gender-related. You will have to treat these two type of words differently when debiasing.

# Neutralize bias for non-gender specific words 

The figure below should help you visualize what neutralizing does. If you're using a 50-dimensional word embedding, the 50 dimensional space can be split into two parts: The bias-direction $g$, and the remaining 49 dimensions, which we'll call $g_{\perp}$. In linear algebra, we say that the 49 dimensional $g_{\perp}$ is perpendicular (or "orthogonal") to $g$, meaning it is at 90 degrees to $g$. The neutralization step takes a vector such as $e_{receptionist}$ and zeros out the component in the direction of $g$, giving us $e_{receptionist}^{debiased}$. 

Even though $g_{\perp}$ is 49 dimensional, given the limitations of what we can draw on a screen, we illustrate it using a 1 dimensional axis below. 


![](images/neutral.png)


<caption><center> <b>Figure 2</b>: The word vector for "receptionist" represented before and after applying the neutralize operation. </center></caption>

**TASK 10**: Implement `neutralize()` to remove the bias of words such as "receptionist" or "scientist". Given an input embedding $e$, you can use the following formulas to compute $e^{debiased}$: 


\begin{equation}
e^{bias\_component} = \frac{e \cdot g}{||g||_2^2} * g\tag{2}
\end{equation}

\begin{equation}
e^{debiased} = e - e^{bias\_component}\tag{3}
\end{equation}

If you are an expert in linear algebra, you may recognize $e^{bias\_component}$ as the projection of $e$ onto the direction $g$. If you're not an expert in linear algebra, don't worry about this.

<!-- 
**Reminder**: a vector $u$ can be split into two parts: its projection over a vector-axis $v_B$ and its projection over the axis orthogonal to $v$:
$$u = u_B + u_{\perp}$$
where : $u_B = $ and $ u_{\perp} = u - u_B $
!--> 
***

In [18]:
# TASK 10
# GRADED neutralize

def neutralize(word, g, word_to_vec_map):
    """
    Removes the bias of "word" by projecting it on the space orthogonal to the bias axis. 
    This function ensures that gender neutral words are zero in the gender subspace.
    
    Arguments:
        word -- string indicating the word to debias
        g -- numpy-array of shape (50,), corresponding to the bias axis (such as gender)
        word_to_vec_map -- dictionary mapping words to their corresponding vectors.
    
    Returns:
        e_debiased -- neutralized word vector representation of the input "word"
        
    """
    g = word_to_vec_map.get('woman') - word_to_vec_map.get('man')
    
    
    e = word_to_vec_map.get(word)
    
    dot_product_g_e = np.dot(g,e)
    
    sum_g = 0
    for j in g:
        j = j ** 2 
        sum_g = sum_g + j

    norm_2_g = math.sqrt(sum_g)
    
    e_bias = (dot_product_g_e/norm_2_g**2) * g
    
    e_debiased = e - e_bias
    
    return e_debiased

    
    ## START YOU CODE HERE

#cosine_similarity_function(word_to_vec_map.get(list_of_words[0]),  word_to_vec_map.get(list_of_words[1]))

    ## END 
    

In [19]:
e = "receptionist"
print("cosine similarity between " + e + " and g, before neutralizing: ", cosine_similarity_function(word_to_vec_map.get("receptionist"), g))

e_debiased = neutralize("receptionist", g, word_to_vec_map)
print("cosine similarity between " + e + " and g, after neutralizing: ", cosine_similarity_function(e_debiased, g))

cosine similarity between receptionist and g, before neutralizing:  0.3307794175059373
cosine similarity between receptionist and g, after neutralizing:  1.5423976002330523e-16


# Equalization algorithm for gender-specific words

Next, lets see how debiasing can also be applied to word pairs such as "actress" and "actor." Equalization is applied to pairs of words that you might want to have differ only through the gender property. As a concrete example, suppose that "actress" is closer to "babysit" than "actor." By applying neutralizing to "babysit" we can reduce the gender-stereotype associated with babysitting. But this still does not guarantee that "actor" and "actress" are equidistant from "babysit." The equalization algorithm takes care of this. 

The key idea behind equalization is to make sure that a particular pair of words are equi-distant from the 49-dimensional $g_\perp$. The equalization step also ensures that the two equalized steps are now the same distance from $e_{receptionist}^{debiased}$, or from any other word that has been neutralized. In pictures, this is how equalization works: 


![](images/equalize10.png)



The derivation of the linear algebra to do this is a bit more complex. (See Bolukbasi et al., 2016 for details.) But the key equations are: 

\begin{equation}
\mu = \frac{e_{w1} + e_{w2}}{2}\tag{4}
\end{equation}

\begin{equation}
\mu_{B} = \frac {\mu \cdot bias\_axis}{||bias\_axis||_2^2} *bias\_axis
\tag{5}
\end{equation}

\begin{equation}
\mu_{\perp} = \mu - \mu_{B} \tag{6}
\end{equation}

\begin{equation}
e_{w1B} = \frac {e_{w1} \cdot bias\_axis}{||bias\_axis||_2^2} *bias\_axis
\tag{7}
\end{equation}

\begin{equation}
e_{w2B} = \frac {e_{w2} \cdot bias\_axis}{||bias\_axis||_2^2} *bias\_axis
\tag{8}
\end{equation}

\begin{equation}
e_{w1B}^{corrected} = \sqrt{ |{1 - ||\mu_{\perp} ||^2_2} |} * \frac{e_{\text{w1B}} - \mu_B} {||(e_{w1} - \mu_{\perp}) - \mu_B||} \tag{9}
\end{equation}

\begin{equation}
e_{w2B}^{corrected} = \sqrt{ |{1 - ||\mu_{\perp} ||^2_2} |} * \frac{e_{\text{w2B}} - \mu_B} {||(e_{w2} - \mu_{\perp}) - \mu_B||} \tag{10}
\end{equation}

\begin{equation}
e_1 = e_{w1B}^{corrected} + \mu_{\perp} \tag{11}
\end{equation}

\begin{equation}
e_2 = e_{w2B}^{corrected} + \mu_{\perp} \tag{12}
\end{equation}

**TASK 11**: Implement the function below. Use the equations above to get the final equalized version of the pair of words. Good luck!

***

In [56]:
# TASK 11
# GRADED equalize


def equalize(pair, bias_axis, word_to_vec_map):
    """
    Debias gender specific words by following the equalize method described in the figure above.
    
    Arguments:
    pair -- pair of strings of gender specific words to debias, e.g. ("actress", "actor") 
    bias_axis -- numpy-array of shape (50,), vector corresponding to the bias axis, e.g. gender
    word_to_vec_map -- dictionary mapping words to their corresponding vectors
    
    Returns
    e_1 -- word vector corresponding to the first word
    e_2 -- word vector corresponding to the second word
    """
    
    w1, w2 = pair
    e_w1, e_w2 = word_to_vec_map.get(w1), word_to_vec_map.get(w2)
    
    mu = (e_w1 + e_w2)/2
    
    sum_bias = 0
    for i in bias_axis:
        i = i ** 2 
        sum_bias = sum_bias + i

    norm_2_bias = math.sqrt(sum_bias)
    
    mu_b = (np.dot(mu,bias_axis)/norm_2_bias**2) * bias_axis
    
    mu_orth = mu - mu_b
    
    e_w1_b = (np.dot(e_w1,bias_axis)/norm_2_bias**2) * bias_axis
    
    e_w2_b = (np.dot(e_w2,bias_axis)/norm_2_bias**2) * bias_axis
    
    sum_mu_orth = 0
    for k in mu_orth:
        k =k ** 2 
        sum_mu_orth = sum_mu_orth + i

    norm_2_mu_orth = math.sqrt(sum_mu_orth)
    
    
    e_w1_b_corrected = math.sqrt(abs(1-(norm_2_mu_orth**2))) * ((e_w1_b - mu_b)/np.linalg.norm(e_w1_b - mu_orth - mu_b,1))
    

    e_w2_b_corrected = math.sqrt(abs(1-(norm_2_mu_orth**2))) * ((e_w2_b - mu_b)/np.linalg.norm(e_w2_b - mu_orth - mu_b,1))
    
    e1 = e_w1_b_corrected + mu_orth
    
    e2 = e_w2_b_corrected + mu_orth
    
    return e1,e2
    
    
    ## START YOU CODE HERE
    

    ##END

In [57]:
print("cosine similarities before equalizing:")
print("cosine_similarity(word_to_vec_map[\"man\"], gender) = ", cosine_similarity_function(word_to_vec_map["man"], g))
print("cosine_similarity(word_to_vec_map[\"woman\"], gender) = ", cosine_similarity_function(word_to_vec_map["woman"], g))
print()
e1, e2 = equalize(("man", "woman"), g, word_to_vec_map)
print("cosine similarities after equalizing:")
print("cosine_similarity(e1, gender) = ", cosine_similarity_function(e1, g))
print("cosine_similarity(e2, gender) = ", cosine_similarity_function(e2, g))

cosine similarities before equalizing:
cosine_similarity(word_to_vec_map["man"], gender) =  -0.1171109576533683
cosine_similarity(word_to_vec_map["woman"], gender) =  0.3566661884627037

cosine similarities after equalizing:
cosine_similarity(e1, gender) =  -0.00934778959133861
cosine_similarity(e2, gender) =  0.009601543890486857


**TASK 12: What do you abserve?**

***

Before equalizing, the cosine similarity of "man" and "woman" with respect to g, which represented gender, was at -0.11711 and 0.35666 respectively. This shows that the word "man" is less bias towards the concept of gender as compared to the word "woman". 

After equalizing was done, the word "man" and "woman" has a value of cosine similarity of -0.0093477 and 0.009601 respectively. This shows that the word "man" and "woman" are almost equidistant with respect to gender, which shows that there are almost no difference in the word "man" and "woman".


***

Please feel free to play with the input words in the cell above, to apply equalization to other pairs of words. 

These debiasing algorithms are very helpful for reducing bias, but are not perfect and do not eliminate all traces of bias. For example, one weakness of this implementation was that the bias direction $g$ was defined using only the pair of words _woman_ and _man_. As discussed earlier, if $g$ were defined by computing $g_1 = e_{woman} - e_{man}$; $g_2 = e_{mother} - e_{father}$; $g_3 = e_{girl} - e_{boy}$; and so on and averaging over them, you would obtain a better estimate of the "gender" dimension in the 50 dimensional word embedding space. Feel free to play with such variants as well.  
                     

# Congratulations!

You've come to the end of this assignment, and have seen a lot of the ways that word vectors can be used as well as  modified. Here are the main points you should remember:

- Cosine similarity a good way to compare similarity between pairs of word vectors. (Though L2 distance works too.) 
- For NLP applications, using a pre-trained set of word vectors from the internet is often a good way to get started. 
- Bias in data is an important problem.
- Neutralize and equalize allow to reduce bias in the data.

Congratulations on finishing this notebook! 



# References
- The debiasing algorithm is from Bolukbasi et al., 2016, [Man is to Computer Programmer as Woman is to
Homemaker? Debiasing Word Embeddings](https://papers.nips.cc/paper/6228-man-is-to-computer-programmer-as-woman-is-to-homemaker-debiasing-word-embeddings.pdf)
- The GloVe word embeddings were due to Jeffrey Pennington, Richard Socher, and Christopher D. Manning. (https://nlp.stanford.edu/projects/glove/)
